src/HPLTokenHandler.mo

```diff
@@ -0,0 +1,331 @@
+import Time = "mo:base/Time";
+import Nat = "mo:base/Nat";
+import Text = "mo:base/Text";
+import R "mo:base/Result";
+
+import CircularBuffer "CircularBuffer";
+import Error "mo:base/Error";
+
+module HPLTokenHandler {
+
+  // https://github.com/research-ag/hpl/#readme
+  public module HPL {
+    public type AssetId = Nat;
+    public type SubaccountId = Nat;
+    public type VirtualAccountId = Nat;
+    public type VirtualAccountUpdateObject = { backingSubaccountId: SubaccountId; assetBalance: Nat };
+    public type Asset = {
+      #ft : (id : AssetId, quantity : Nat);
+    };
+    public type VirtualAccountState = { asset: Asset; backingSubaccountId: SubaccountId; remotePrincipal: Principal };
+    public type TxInput = {
+      map : [ContributionInput];
+    };
+    public type AccountReference = {
+      #sub : SubaccountId;
+      #vir : (Principal, VirtualAccountId);
+    };
+    type ContributionBody = {
+      inflow : [(AccountReference, Asset)];
+      outflow : [(AccountReference, Asset)];
+      mints : [Asset];
+      burns : [Asset];
+      memo : ?Blob;
+    };
+    public type ContributionInput = ContributionBody and {
+      owner : ?Principal;
+    };
+    public type AggregatorId = Nat;
+    public type LocalId = Nat;
+    public type GlobalId = (aggregator : AggregatorId, localId : LocalId);
+    public type ProcessingError = {
+      #TooLargeAssetId;
+      #TooLargeFtQuantity;
+      #TooLargeSubaccountId;
+      #TooLargeVirtualAccountId;
+      #TooLargeMemo;
+      #TooManyFlows;
+      #TooManyContributions;
+      #NonZeroAssetSum;
+      #UnknownPrincipal;
+      #UnknownSubaccount;
+      #UnknownVirtualAccount;
+      #DeletedVirtualAccount;
+      #UnknownFtAsset;
+      #MismatchInAsset;
+      #MismatchInRemotePrincipal;
+      #InsufficientFunds;
+      #NotAController;
+    };
+    public type SubmitAndExecuteError = ProcessingError or { #NotApproved; };
+    public type Ledger = actor {
+      openVirtualAccount : (state : VirtualAccountState) -> async R.Result<VirtualAccountId, { #UnknownPrincipal; #UnknownSubaccount; #MismatchInAsset; #NoSpaceForAccount }>;
+      setVirtualBalance : (vid : VirtualAccountId, newBalance : Nat) -> async R.Result<Int, { #UnknownPrincipal; #UnknownVirtualAccount; #DeletedVirtualAccount }>;
+      incVirtualBalance : (vid : VirtualAccountId, delta : Nat) -> async R.Result<Nat, { #UnknownPrincipal; #UnknownVirtualAccount; #DeletedVirtualAccount }>;
+      decVirtualBalance : (vid : VirtualAccountId, delta : Nat) -> async R.Result<Nat, { #InsufficientFunds; #UnknownPrincipal; #UnknownVirtualAccount; #DeletedVirtualAccount }>;
+      virtualAccount : (vid : VirtualAccountId) -> async R.Result<VirtualAccountState, { #UnknownPrincipal; #UnknownVirtualAccount; #DeletedVirtualAccount }>;
+      submitAndExecute : (tx : TxInput) -> async R.Result<GlobalId, SubmitAndExecuteError>;
+    };
+  };
+
+  public func defaultHandlerStableData(): StableData = ([], 0, (0, 0), ([var], 0, 0));
```

Comment:
Is this function used yet?

Reply:
It should be used by parent application as a shortcut for declaring stable variable, like:
`stable var handlerData: HPLTokenHandler.StableData = HPLTokenHandler.defaultHandlerStableData();`